File: app/models/Jogo.scala
package models

import java.sql.Timestamp
import java.text.{SimpleDateFormat, DateFormat}
import java.util.Date

import org.joda.time.DateTime
import org.joda.time.format.DateTimeFormat
import play.api.Play.current
import play.api.db.slick.Config.driver.simple._
import scala.slick.lifted.Tag

object date {
  val df = DateTimeFormat.forPattern("DDMM")

  def capaData(d: DateTime) = {
    df.print(d);
  }
}
case class Jogo(id: Option[Long], titulo: String, capa: String,var data: DateTime,  var disponibilidade: Boolean) {

  def capaNome = if (disponibilidade) capa else "ua-" + capa
  def capaData = date.capaData(data)
}

class Jogos(tag: Tag) extends Table[Jogo](tag, "JOGOS") {
  implicit def DateTime2TimeStamp = MappedColumnType.base[DateTime, Timestamp]({
    dt => new Timestamp(dt.getMillis )
  }, {
    tt => new DateTime(tt)
  })

  def id = column[Option[Long]]("id", O.PrimaryKey, O.AutoInc)
  def titulo = column[String]("titulo", O.NotNull)
  def capa = column[String]("capa", O.NotNull)
  def data = column[DateTime]("data")
  def disponibilidade = column[Boolean]("disponibilidade", O.NotNull)

  override def * = (id, titulo, capa, data, disponibilidade) <> (Jogo.tupled, Jogo.unapply _)
}


object Jogos {
  val jogos = TableQuery[Jogos]

  def list(implicit s: Session) = {
    val query = jogos.list
    query toList
  }

  def get(id: Long)(implicit s: Session): Jogo = {
    jogos.filter( _.id === id ) first
  }

  def insert(jogo: Jogo)(implicit s: Session) {
    jogos.insert(jogo)
  }

  def update(id: Long, jogo: Jogo)(implicit s: Session): Boolean = {
    jogos.filter(_.id === id ).update(jogo) > 0
  }
}

File: app/models/Jogo.java
package models;

import com.google.common.base.Objects;

import javax.persistence.*;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;

@Entity
@Table(name = "JOGOS")
public class Jogo implements Cloneable {

    @Id
    @GeneratedValue(strategy= GenerationType.SEQUENCE)
    public Long id;
    public final String titulo;
    public final String capa;
    public Boolean disponibilidade;

    @Temporal(TemporalType.TIMESTAMP)
    public Date data_reserva;

    @Temporal(TemporalType.TIMESTAMP)
    public final Date created_on = new Date();

    public Jogo() {
        this("", "",new Date(), false);
    }

    public Jogo(String titulo, String capa, Date data_reserva, boolean disponiblidade){
        this.titulo = titulo;
        this.capa = capa;
        this.data_reserva = data_reserva;
        this.disponibilidade = disponiblidade;
    }

    public String getDataReservaFormatada(){
        DateFormat df = new SimpleDateFormat("yyyy-MM-dd");
        return df.format(this.data_reserva);
    }

    @Override
    public Jogo clone() throws CloneNotSupportedException {
        return (Jogo) super.clone();
    }

    @Override
    public String toString() {
        return "Jogo{" +
                "id=" + id +
                ", titulo='" + titulo + '\'' +
                ", cover='" + capa + '\'' +
                ", disponibilidade=" + disponibilidade +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Jogo jogo = (Jogo) o;
        return Objects.equal(id, jogo.id) &&
                Objects.equal(titulo, jogo.titulo) &&
                Objects.equal(capa, jogo.capa) &&
                Objects.equal(disponibilidade, jogo.disponibilidade);
    }

    @Override
    public int hashCode() {
        return Objects.hashCode(id, titulo);
    }

    public Jogo copy(Long id) {
        try {
            Jogo copy = clone();
            copy.id = id;
            return copy;
        }catch (CloneNotSupportedException e){
            return null;
        }
    }
}
